agent/transport/transport.go

```diff
@@ -0,0 +1,156 @@
+package transport
+
+import (
+	"io"
+	"sort"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+// A Factory is used to initialize a new transport.
+type Factory func(*zap.Logger, TaskWriter) (io.WriteCloser, error)
+
+// A TaskWriter is provided to a Transport to enable it to send incoming tasks to the agent.
+type TaskWriter interface {
+	WriteTask(id string, content io.Reader)
+}
+
+// An Option enables additional configuration of a transport Meta.
+type Option func(*Meta)
+
+// Meta holds metadata about a transport and a factory method to initialize it.
+type Meta struct {
+	Name     string
+	Priority int
+	Interval time.Duration
+	Jitter   time.Duration
+	Factory  Factory
+}
+
+// A Registry maintains a mapping of transport metadata registered with it and provides functionality
+// to return a sorted list of registered transport metadata.
+type Registry struct {
+	SortBy func(t1, t2 Meta) bool
+
+	active   map[string]io.WriteCloser
+	metadata map[string]*Meta
+}
+
+// Register a transport factory and configure it's metadata.
+func (reg Registry) Register(name string, factory Factory, options ...Option) {
+	meta := &Meta{
+		Name:    name,
+		Factory: factory,
+	}
+
+	for _, opt := range options {
+		opt(meta)
+	}
+
+	reg.metadata[name] = meta
+}
+
+// Update transport metadata by applying options to the transport metadata with the provided name.
+func (reg Registry) Update(name string, options ...Option) error {
+	transport, ok := reg.metadata[name]
+	if !ok {
+		return errors.New("transport not registered")
+	}
+
+	for _, opt := range options {
+		opt(transport)
+	}
+	return nil
+}
+
+// Get returns a named transport io.WriteCloser.
+func (reg Registry) Get(name string, logger *zap.Logger, writer TaskWriter) (io.WriteCloser, error) {
+	transport, ok := reg.active[name]
+	if ok && transport != nil {
+		return transport, nil
+	}
+
+	meta, ok := reg.metadata[name]
+	if !ok || meta == nil || meta.Factory == nil {
```

Comment:
what is this doing?

Reply:
Get the metadata for the transport with the provided name from the registry. If the transport is not registered (`!ok`), it is a nil pointer (`meta == nil`), or has a nil factory (`meta.Factory ==nil`), then remove it from the registry and return an error saying the transport is not registered.